### PR TITLE
remove dnsLookUp func from traffic resources

### PR DESCRIPTION
**Pull request by maleck13**
Closes #

## Description of Changes
The traffic resources, as much as possible should focus on mutating the structure of the underlying resource. This PR removes the need to pass in a dnsLookUp function and simplifies the GetDNS() function considerably. This function was actually redundant due to the fact that any host value LB are added to the host watcher. Additionally this adds some new unit tests to cover the changed funcitons

## Remaining Work 

The following changes still need to be completed:

- [ ] List any outstanding work here

## Release and Testing

*

## Dependencies

- [ ] Perequisites that must be completed before merging this pull request

### CODE REVIEW

File: pkg/dns/target.go [hunk 1]
```diff
@@ -6,6 +6,8 @@ const (
 )
 
 type Target struct {
+	Host       string
```
Comment (mikenairn) [inline pkg/dns/target.go:9]: Ha, you've made changes that i was also making :+1:

File: pkg/traffic/dns.go [hunk 2]
```diff
@@ -99,28 +98,24 @@ func (r *DnsReconciler) Reconcile(ctx context.Context, accessor Interface) (Reco
...
-	targets, err := accessor.GetDNSTargets(ctx, r.DNSLookup)
+	targets, err := accessor.GetDNSTargets()
 	if err != nil {
 		return ReconcileStatusContinue, err
 	}
-	for cluster, targets := range targets {
-		deleteAnnotation := workload.InternalClusterDeletionTimestampAnnotationPrefix + cluster.String()
+
+	for _, target := range targets {
+		host := target.Value
```
Comment (mikenairn) [inline pkg/traffic/dns.go:107]: Is `target.Host` used anywhere? Seems odd to create a local var called `host` that doesn't use `target.Host`

File: pkg/dns/target.go [hunk 3]
```diff
@@ -6,6 +6,8 @@ const (
...
+	Cluster    string
 	TargetType string
-	Value      []string
+	Value      string
```
Comment (mikenairn) [inline pkg/dns/target.go:12]: Are multiple targets created for hosts with more than a single IP? ELBs for instance?

File: pkg/traffic/dns.go (hunk 2 above)
Comment (mikenairn) [reply pkg/traffic/dns.go:107]: Can `target.Value` be a hostname or an IP now? Previously it was just an array of IPs right?

File: pkg/dns/target.go (hunk 3 above)
Comment (maleck13) [reply pkg/dns/target.go:12]: Yes 1 target per host or per IP

Comment (maleck13) [reply pkg/dns/target.go:12]: however I think there might be a bug you just highlighted. There is a host watcher that looks up IPs for a host. I had thought it was updating the DNSRecord but I think it only triggers a new reconcile of the ingress, so will check into that

Comment (maleck13) [reply pkg/dns/target.go:12]: Yes so the main issue we have is none of our tests cover when there is a host set other than our smoke test. I will add an issue to correct this.

File: pkg/traffic/dns.go (hunk 2 above)
Comment (maleck13) [reply pkg/traffic/dns.go:100]: yes now there is a dns target per Ip or Host and value can be either so I have removed the Host field now

**Comment by maleck13** [comment]
@mikenairn this is ready for a final review now. I have added a unit test to cover the case where there is a host rather than an IP

File: pkg/traffic/dns.go [hunk 4]
```diff
@@ -95,31 +86,45 @@ func (r *DnsReconciler) Reconcile(ctx context.Context, accessor Interface) (Reco
...
+		if err != nil {
+			return ReconcileStatusContinue, fmt.Errorf("DNSLookup failed for host %s : %s", host, err)
+		}
+		for _, add := range addr {
+			activeDNSTargetIPs[host] = append(activeDNSTargetIPs[host], add.IP.String())
+		}
+		//add the host to host watcher to keep our DNS upto date
+		// If it is not an IP we add it to the host watcher that triggers an update when it gets IPS
+		r.WatchHost(ctx, key, host)
+		activeLBHosts = append(activeLBHosts, host)
```
Comment (mikenairn) [inline pkg/traffic/dns.go:120]: Why do we no do this for both IPs and Hosts?

Comment (mikenairn) [reply pkg/traffic/dns.go:120]: nvm, we are continuing above if we got an IP

File: pkg/traffic/dns.go [hunk 5]
```diff
@@ -95,31 +86,45 @@ func (r *DnsReconciler) Reconcile(ctx context.Context, accessor Interface) (Reco
...
-				r.WatchHost(ctx, key, host)
-				activeLBHosts = append(activeLBHosts, host)
-			}
-			activeDNSTargetIPs[host] = append(activeDNSTargetIPs[host], target.Value...)
+		if target.TargetType == dns.TargetTypeIP {
+			activeDNSTargetIPs[host] = append(activeDNSTargetIPs[host], host)
+			continue
+		}
+		// for a non ip value look up the DNS
+		addr, err := r.DNSLookup(ctx, target.Value)
```
Comment (mikenairn) [inline pkg/traffic/dns.go:110]: Should this pass in `host` instead of `target.Value` here?

File: pkg/traffic/route.go [hunk 6]
```diff
@@ -118,31 +118,20 @@ func (a *Route) RemoveTLS(hosts []string) {
 	}
 }
 
-func (a *Route) GetDNSTargets(ctx context.Context, dnsLookup dnsLookupFunc) (map[logicalcluster.Name]map[string]dns.Target, error) {
-	targets := map[logicalcluster.Name]map[string]dns.Target{}
+func (a *Route) GetDNSTargets() ([]dns.Target, error) {
+	targets := []dns.Target{}
```
Comment (mikenairn) [inline pkg/traffic/route.go:122]: nit: Should probably rename this `dnsTargets` for consistency with the ingress changes

**Review by mikenairn — APPROVED** [review]
Couple of small things, but nothing that need to stop this merging.

**Pull request merged**